lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -447,9 +447,13 @@ void SPIRVRegularizeLLVMBase::adaptStructTypes(StructType *ST) {
            "Expected a pointer to an array to represent joint matrix type");
     size_t TypeLayout[4] = {0, 0, 0, 0};
     ArrayType *ArrayTy = dyn_cast<ArrayType>(PtrTy->getPointerElementType());
+    assert(ArrayTy && "Expected a pointer element type of an array type to "
+                      "represent joint matrix type");
```

Comment:
If we are not checking an `ArrayTy` for `nullptr`, then we should use simple `cast` to avoid runtime `isa` check within `dyn_cast`

Reply:
Maybe the intention of using `dyn_cast` was to assert with a more meaningful message than `cast` would raise. And actually `cast ` does `isa` check, and asserts if it failed - `dyn_cast` seems to be more gentle in this sense. 
But agree that it's not the general use for `dyn_cast`. Tagging @MrSidims as an author to take a look.

Reply:
Keep in mind that the `getPointerElementType` call eventually has to be removed: the code cannot rely on obtaining the element type through `PtrTy` once LLVM has transitioned fully to opaque pointers, see https://llvm.org/docs/OpaquePointers.html for more info.  It might be worth trying to remove the `getPointerElementType` call now if anyone is going to look at improving this code.

Reply:
> And actually `cast` does `isa` check

It does so only within an `assert` meaning that it has no overhead on a runtime check. If we anyway ignore the result of a runtime check performed by `dyn_cast`, then why do it in the first place? I tend to think that you should be able to quickly understand what went wrong even with generic assert message from within `cast` by simply looking at backtrace and/or moving a few frames upwards in a debugger.